aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,367 @@
+    payload_.boxed_iterator = boxed;
+  }
+
+  ITensorListIterator(unboxed_iterator_t unboxed)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed_iterator = unboxed;
+  }
+
+  detail::ITensorListConstRef operator*() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return ImplT::iterator_get(this_); });
+  }
+
+  ITensorListIterator& operator++() {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { ++this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator++(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { ++this_; });
+    return old;
+  }
+
+  ITensorListIterator& operator--() {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator--(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return old;
+  }
+
+  bool operator==(const ITensorListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ITENSORLIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const ITensorListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  Payload payload_;
+  ITensorListTag tag_;
+};
+
+/*
+ * Wrapper around boxed and unboxed API containers.
+ *
+ * Tagged union of both API containers:
+ *     - `TensorList`, a.k.a. `ArrayRef<Tensor>` (the unboxed API container)
+ *     - `List<Tensor>` (the boxed API container)
+ *
+ * This container wraps around these two, without incurring in extra overhead
+ * for converting from one to another.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ */
+class ITensorList {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...) friend class TORCH_ITENSORLIST_IMPL(TAG);
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_FRIEND_CLASS);
+#undef DEFINE_FRIEND_CLASS
+
+  using unboxed_t = TORCH_ITENSORLIST_IMPL(Unboxed)::self_t;
+  using boxed_t = TORCH_ITENSORLIST_IMPL(Boxed)::self_t;
+
+  union Payload {
+    const boxed_t* boxed;
+    unboxed_t unboxed;
+    Payload() : boxed(nullptr) {}
+    ~Payload() = default;
+  };
+
+ public:
+  using iterator = ITensorListIterator;
+  using const_iterator = ITensorListIterator;
+  using value_type = typename iterator::value_type;
+
+  ITensorList() : tag_(ITensorListTag::None) {}
+
+  ITensorList(const std::initializer_list<at::Tensor>& list)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = at::ArrayRef<at::Tensor>(list);
+  }
+
+  ITensorList(const boxed_t& boxed) : tag_(ITensorListTag::Boxed) {
+    payload_.boxed = &boxed;
+  }
+
+  ITensorList(const unboxed_t& unboxed) : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = unboxed;
+  }
+
+  template <
+      typename... UnboxedConstructorArgs,
+      typename = std::enable_if_t<
+          std::is_constructible<unboxed_t, UnboxedConstructorArgs...>::value>>
+  ITensorList(UnboxedConstructorArgs&&... args)
```

Comment:
Hmm, is the idea that this constructor will let you do something like this:
```
TORCH_API inline at::Tensor cat(at::ITensorList tensors, int64_t dim=0)
...
at::cat(tensor1, tensor2)   # under the hood we'll back this by an `ArrayRef<Tensor>`
```

That sounds to me like new behavior. Is that currently used/needed anywhere? I'm not totally sure that we want to allow `ITensorList` to be implicitly constructable out of anything that you can't already implicitly construct a `c10::ArrayRef` or `c10::List` with.

We would like initializer lists to work:
```
at::cat({tensor1, tensor2})
```
But I think we should get this for free, right? Since you can implicitly construct an initializer list -> ArrayRef, and ArrayRef -> ITensorList.

Reply:
> Hmm, is the idea that this constructor will let you do something like this:

I think that `at::cat(tensor1, tensor2)` wouldn't compile... Mostly, this constructor enables us to use `c10::ArrayRef` constructors that take more than 1 parameter

https://github.com/pytorch/pytorch/blob/5fdcc20d8d96a6b42387f57c2ce331516ad94228/c10/util/ArrayRef.h#L74-L84

> I'm not totally sure that we want to allow `ITensorList` to be implicitly constructable out of anything that you can't already implicitly construct a `c10::ArrayRef` or `c10::List` with.

I'm doing that only for the unboxed `c10::ArrayRef`, since the idea is that we don't own the boxed container. Since the `c10::ArrayRef` is only a "view" of a container-like thing, I thougth it would be a good idea to be able to do it directly with `ITensorList`.

> But I think we should get this for free, right? Since you can implicitly construct an initializer list -> ArrayRef, and ArrayRef -> ITensorList.

Yes. This one we get for free.

Reply:
> I don't think I'm following. If we remove this constructor, what functionality do we lose?
> 
> We still want `at::cat({list.begin(), list.end()})` to work, but I don't think we need that templated constructor to get it - which has the downside that explicit `ArrayRef` constructors will be implicit for `ITensorList`.

Ah, you are right. We need that so that `ITensorList(list.begin(), list.end())` also works. In other words, that makes all `ArrayRef` constructors available for `ITensorList`, when we want to use it explicitly. But, you are right, maybe we don't really want that.

Reply:
@bdhirsh
After thinking about it for a while, I realized that what happens is a little bit different than what we were discussing here.
Check out [this godbolt link](https://godbolt.org/z/W8ocxG86q).

In summary, without that constructor we **can't** execute `at::cat({list.begin(), list.end()})`. We **can**, however, execute `at::cat({{list.begin(), list.end()}})` (note the double braces). Because of that, I think we'd want to keep this constructor.

Let me know what you think.

Reply:
Hmm ok, I'm still not 100% convinced that this constructor is the only way to fix the problem, but it seems low risk to add.

Reply:
Well, I guess we could just replicate the constructors from `ArrayRef` that we want for the sake of being explicit, with the downside that we'll be a little bit more verbose. Happy to take any suggestions!

Reply:
The template here seems fine.